Objects/typeobject.c

```diff
@@ -6665,47 +6671,48 @@ object_set_class(PyObject *self, PyObject *value, void *closure)
         if (oldto->tp_flags & Py_TPFLAGS_INLINE_VALUES) {
             PyDictObject *dict = _PyObject_MaterializeManagedDict(self);
             if (dict == NULL) {
-                return -1;
+                goto err;
             }
 
-            bool error = false;
-
-            Py_BEGIN_CRITICAL_SECTION2(self, dict);
-
             // If we raced after materialization and replaced the dict
             // then the materialized dict should no longer have the
             // inline values in which case detach is a nop.
+            // Note: we don't need to lock here because the world should be stopped.
+
             assert(_PyObject_GetManagedDict(self) == dict ||
                    dict->ma_values != _PyObject_InlineValues(self));
 
             if (_PyDict_DetachFromObject(dict, self) < 0) {
-                error = true;
+                goto err;
             }
 
-            Py_END_CRITICAL_SECTION2();
-            if (error) {
-                return -1;
-            }
         }
         if (newto->tp_flags & Py_TPFLAGS_HEAPTYPE) {
             Py_INCREF(newto);
         }
-        Py_BEGIN_CRITICAL_SECTION(self);
-        // The real Py_TYPE(self) (`oldto`) may have changed from
-        // underneath us in another thread, so we re-fetch it here.
+
         oldto = Py_TYPE(self);
         Py_SET_TYPE(self, newto);
-        Py_END_CRITICAL_SECTION();
+
         if (oldto->tp_flags & Py_TPFLAGS_HEAPTYPE) {
             Py_DECREF(oldto);
```

Comment:
The `Py_DECREF()` should be after the `_PyEval_StartTheWorld()` because it may call arbitrary code through a finalizer.